File: cli/output.py
from __future__ import annotations

import json

from rich.console import Console

console = Console()


def print_result(payload: dict, output_mode: str = "human") -> None:
    if output_mode == "json":
        console.print(json.dumps(payload, indent=2, ensure_ascii=True))
        return

    handled: set[str] = set()
    status = payload.get("status")
    if status is not None:
        handled.add("status")
    if status == "ok":
        console.print("[green]ok[/green]")
    elif status is not None:
        console.print(f"[red]{status}[/red]")

    if "response" in payload:
        handled.add("response")
        console.print(payload["response"])

    if "error" in payload:
        handled.add("error")
        console.print(f"[red]error:[/red] {payload['error']}")

    run = payload.get("run")
    if isinstance(run, dict):
        handled.add("run")
        run_id = run.get("id", "<unknown>")
        run_status = run.get("status", "unknown")
        completed = run.get("completed_steps")
        total = run.get("total_steps")
        failure_reason = run.get("failure_reason")
        progress = ""
        if completed is not None and total is not None:
            progress = f" ({completed}/{total})"
        if failure_reason:
            console.print(f"run {run_id}: {run_status}{progress}, reason={failure_reason}")
        else:
            console.print(f"run {run_id}: {run_status}{progress}")

    remaining = {k: v for k, v in payload.items() if k not in handled}
    if remaining:
        console.print(json.dumps(remaining, indent=2, ensure_ascii=True))

File: earl-cli/src/output.js
function stableStringify(obj) {
  return JSON.stringify(obj, null, 2);
}

function printResult(payload, outputMode = "human") {
  const mode = String(outputMode || "human").toLowerCase() === "json" ? "json" : "human";
  const data = payload && typeof payload === "object" ? payload : { status: "error", error: "Invalid payload" };

  if (mode === "json") {
    // eslint-disable-next-line no-console
    console.log(stableStringify(data));
    return;
  }

  const handled = new Set();
  const status = data.status;
  if (status != null) handled.add("status");
  if (status === "ok") {
    // eslint-disable-next-line no-console
    console.log("ok");
  } else if (status != null) {
    // eslint-disable-next-line no-console
    console.log(String(status));
  }

  if (Object.prototype.hasOwnProperty.call(data, "response")) {
    handled.add("response");
    // eslint-disable-next-line no-console
    console.log(String(data.response));
  }

  if (Object.prototype.hasOwnProperty.call(data, "error")) {
    handled.add("error");
    // eslint-disable-next-line no-console
    console.log(`error: ${String(data.error)}`);
  }

  const run = data.run;
  if (run && typeof run === "object" && !Array.isArray(run)) {
    handled.add("run");
    const runId = run.id || "<unknown>";
    const runStatus = run.status || "unknown";
    const completed = run.completed_steps;
    const total = run.total_steps;
    const failureReason = run.failure_reason;
    let progress = "";
    if (completed != null && total != null) progress = ` (${completed}/${total})`;
    // eslint-disable-next-line no-console
    console.log(
      failureReason
        ? `run ${runId}: ${runStatus}${progress}, reason=${failureReason}`
        : `run ${runId}: ${runStatus}${progress}`
    );
  }

  const remaining = {};
  for (const [k, v] of Object.entries(data)) {
    if (!handled.has(k)) remaining[k] = v;
  }
  if (Object.keys(remaining).length > 0) {
    // eslint-disable-next-line no-console
    console.log(stableStringify(remaining));
  }
}

module.exports = { printResult };

